Updates for capgen-v1 (#81)
NCAR/ccpp-doc: CCPPtechnical/source/CCPPPreBuild.rst

.. _CCPPCapgen:

**************************************************
Technical Aspects of CCPP *Capgen*
**************************************************

:term:`CCPP` *capgen* consists of three primary scripts:

* ccpp_validator.py           - Compare metadata files against Fortran source files

* ccpp_capgen.py              - Autogenerate physics caps

* ccpp_datafile.py            - Query information on Capgen files, variables, and groups

.. _ccpp_validator:

=============================
*ccpp_validator.py*
=============================

The :term:`CCPP` *validator* script ``ccpp-framework/capgen/ccpp_validator.py`` is used to validate the metadata files against the accompanying Fortran source code.
This script can (should) be invoked prior to calling the primary ``ccpp_capgen.py`` script to ensure CCPP compliance, and to suss out any inconsistencies at build-time rather than run-time.

.. code-block:: console

  --host-files                 - Comma-separated list of host-model metadata (.meta) files

>   --scheme-files               - Comma-separated list of scheme scheme metadata (.meta) files

Reviewer: scheme-files is mutually exclusive with host-files; ccpp_validator can only validate one type of metadata (either schemes or host) against a set of Fortran files

I think this section would benefit from a minimal example, e.g. one of the basic end-to-end tests in the ccpp-framework repository

Reviewer: nevermind, I saw that you did that further down

Author: --host-files argument removed.

  --source-files               - Comma-separated Fortran source (.F90) files

*Listing 8.1: Options for* ``ccpp_validator.py`` *script.*

> The ``ccpp_validator.py`` script can be used to validate both *Scheme files* and *Host files*. 

Reviewer: ```suggestion
The ``ccpp_validator.py`` script can be used to validate either *scheme files* or *host files*. 
```

=============================
*ccpp_capgen.py*
=============================

The :term:`CCPP` *capgen* script ``ccpp-framework/capgen/ccpp_capgen.py`` is the central piece of code in the :term:`CCPP Framework` that
connects the :term:`host model` with the :term:`CCPP Physics` schemes (see :numref:`%s <ccpp_static_build>`). This script must be run
before compiling the CCPP Physics library and the host model driver. This may be done manually or as part
of a host model build-time script. Both the :term:`UFS` and :term:`SCM` have incorporated the calls to ``ccpp_capgen.py`` in their build systems.

.. code-block:: console
 
  --host-name                  - Host model identifier.  Drives the file and module name of the generated static-API cap
  --host-files                 - Comma-separated list of host-model metadata (.meta) files
  --scheme-files               - Comma-separated list of physics scheme metadata (.meta) files
  --suites                     - Comma-separated list of suite XML definition (.xml) files
  --output-root                - Output directory for all generated files
  --kind-type                  - Map a CCPP kind name to a Fortran precision constant

>   --verbose                    - Increase verbosity.  Use once for INFO messages, twice (-vv) for DEBUG messages

Reviewer: default logging level is WARNING

> *Listing 8.3: Options for* ``ccpp_capgen.py`` *script.*

Reviewer: this is a repeat of AutoGenPhysCaps.rst. we should only list these files in one place and point to it frmo the other place

Author: done.

> The *ccpp_capgen* script reads host-model metadata files, physics scheme metadata files, and suite definition (XML) files, resolves all variable connections and writes:

Reviewer: We should mention that ccpp_capgen only returns a status code, nothing else. Any interaction with the build artifacts happens through ccpp_datafile and the datatable.xml generated by ccpp_capgen.

.. code-block:: console

  ccpp_kinds.F90               — Kind parameter definitions
  <host>_ccpp_cap.F90          — Static dispatch API (per-host; filename and module name derived from ``--host-name``)
  ccpp_<suite>_cap.F90         — Suite-level cap (state machine, group dispatch)
  ccpp_<suite>_<group>_cap.F90 — Group-level cap (scheme call sites)
  ccpp_<suite>_data.F90        — Suite-owned interstitial data module
  ccpp_<suite>_types.F90       — Shared types (pointer wrappers, temp locals)

>   ccpp_<suite>.meta            — Generated suite metadata (for inspection)

Reviewer: ```suggestion
  ccpp_<suite>_data.meta            — Generated suite metadata (for inspection)
```

  # Scheme files  
  set(SCHEME_METADATA_FILES 
      "${CMAKE_SOURCE_DIR}/ccpp/physics/Radiation/RRTMGP/rrtmgp_sw_main.meta"
      "${CMAKE_SOURCE_DIR}/ccpp/physics/Radiation/RRTMGP/rrtmgp_lw_main.meta")
  set(SCHEME_FORTRAN_FILES 
      "${CMAKE_SOURCE_DIR}/ccpp/physics/Radiation/RRTMGP/rrtmgp_sw_main.F90"
      "${CMAKE_SOURCE_DIR}/ccpp/physics/Radiation/RRTMGP/rrtmgp_lw_main.F90")

  # Host files
  set(HOST_METADATA_FILES
      "${CMAKE_SOURCE_DIR}/ccpp/data/CCPP_typedefs.meta"
      "${CMAKE_SOURCE_DIR}/ccpp/data/GFS_typedefs.meta")
  set(HOST_FORTRAN_FILES
      "${CMAKE_SOURCE_DIR}/ccpp/data/CCPP_typedefs.F90"
      "${CMAKE_SOURCE_DIR}/ccpp/data/GFS_typedefs.F90")

  # Suite Definition File(s)
  set(SUITE_FILES 
      "${CMAKE_SOURCE_DIR}/ccpp/suites/suite_FV3_GFS_v17_p8_rrtmgp.xml")

  # Run ccpp_validator for Scheme files
  ${CMAKE_SOURCE_DIR}/ccpp-framework/capgen/ccpp_validator.py --source_files ${SCHEME_FORTRAN_FILES} --scheme-files ${SCHEME_METADATA_FILES} 

  # Run ccpp_validator.py for Host files/
  ${CMAKE_SOURCE_DIR}/ccpp-framework/capgen/ccpp_validator.py --source_files ${HOST_FORTRAN_FILES} --scheme-files ${HOST_METADATA_FILES}

  # Run ccpp_capgen.py

>   ${CMAKE_SOURCE_DIR}/ccpp-framework/capgen/ccpp_capgen.py  --host-files   ${HOST_METADATA_FILES}

Reviewer: Should we mention the optional TRACE feature (which prints the trace log at the top of every auto-generated function in the caps?

Author: Good idea. Added description of ---trace to ccpp_capgen.py above.

                                                            --suites       ${SUITE_FILES}
                                                            --host-name    XYZXYZXYZ
                                                            --output-root  ${CMAKE_CURRENT_BINARY_DIR}/ccpp

  # Run ccpp_datafile.py to create file lists for build.
  ${CMAKE_SOURCE_DIR}/ccpp-framework/capgen/ccpp_datafile.py --capgen-files ${CAPGEN_FILES}
  ${CMAKE_SOURCE_DIR}/ccpp-framework/capgen/ccpp_datafile.py --scheme-files ${SCHEME_FORTRAN_FILES}
  ${CMAKE_SOURCE_DIR}/ccpp-framework/capgen/ccpp_datafile.py --dependencies ${DEPENDENCY_FILES}

  # Build CCPP library
  set(CCPP_TARGET ccpp_capgen_example)
  add_library(${CCPP_TARGET}
              ${DEPENDENCY_FILES}
              ${SCHEME_FORTRAN_FILES}
              ${HOST_FORTRAN_FILES}
              ${CAPGEN_FILES}

  # Install
  install(TARGETS ${CCPP_TARGET}
          EXPORT ${CCPP_TARGET}-config
          LIBRARY DESTINATION lib
          ARCHIVE DESTINATION lib)

*Listing 8.7: Example CMakeLists.txt file for calling the three Capgen entry points.

=============================
Troubleshooting
=============================

> **NEED TO CREATE COMMON CAPGEN ERRORS AND SOLUTIONS**

Reviewer: We can ask Claude to generate a comprehensive list of these errors and solutions

========================================================
CCPP Physics Variable Tracker
========================================================

> **DOES THIS STILL WORK? OR IS THERE A CAPGEN EQUIVALENT?**

Reviewer: Not yet. Same for the debugging routines (formally GFS_debug.F90).

Reviewer: See https://github.com/NCAR/ccpp-framework/issues/770
